// vanil/dist/src/core/loader/core/mdvdom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerMdVdomLoader = exports.mdVdomLoader = exports.markdownToVdom = void 0;
const file_1 = require("../../io/file");
const transpile_1 = require("../../transform/transpile");
const marked = require('marked');
/** transforms markdown to VDOM */
const markdownToVdom = (md, context) => {
    const code = (0, transpile_1.escapeCurlyBracketsAndBackticks)(marked.parse(md));
    // transpiles the SVG content as TSX to functional calls
    // and evaluates them down to a JSX JSON tree representation
    // that can be rendered at runtime
    // TODO: vm -> run()
    return eval((0, transpile_1.transpileTSX)(`const { Code, Debug, Trans } = Vanil;(<>${code}</>)`, context));
};
exports.markdownToVdom = markdownToVdom;
/** resolves a given target path to an absolute path on disk and returns its UTF8 contents */
const mdVdomLoader = (targetPath, context) => (0, exports.markdownToVdom)((0, file_1.readFileSyncUtf8)(targetPath), context);
exports.mdVdomLoader = mdVdomLoader;
const registerMdVdomLoader = () => ({
    name: 'md-vdom',
    match: /\.md$/i,
    cb: exports.mdVdomLoader,
});
exports.registerMdVdomLoader = registerMdVdomLoader;
