// vanil/dist/src/core/optimize/development/prettify.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prettify = void 0;
const prettier_1 = __importDefault(require("prettier"));
const prettify = (code) => {
    try {
        return prettier_1.default.format(code, {
            parser: 'html',
            arrowParens: 'always',
            bracketSameLine: true,
            bracketSpacing: true,
            embeddedLanguageFormatting: 'auto',
            htmlWhitespaceSensitivity: 'css',
            insertPragma: false,
            jsxSingleQuote: false,
            proseWrap: 'preserve',
            quoteProps: 'as-needed',
            requirePragma: false,
            semi: false,
            singleQuote: true,
            tabWidth: 2,
            trailingComma: 'es5',
            useTabs: false,
            vueIndentScriptAndStyle: false,
            printWidth: 120,
        });
    }
    catch (e) { }
    return code;
};
exports.prettify = prettify;
