// vanil/dist/src/core/config/defaults.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONFIG_DEVOPTIONS_USEOPTIMIZER = exports.CONFIG_DEVOPTIONS_USETLS = exports.CONFIG_DEVOPTIONS_PORT = exports.CONFIG_DEVOPTIONS_HOSTNAME = exports.CONFIG_BUILD_PAGE_URL_FORMAT = exports.CONFIG_BUILD_TARGET = exports.CONFIG_BUILD_ROBOTS_TXT = exports.CONFIG_BUILD_SITEMAP = exports.CONFIG_HOOKS_FOLDER = exports.CONFIG_PUBLIC_FOLDER = exports.CONFIG_PAGES_FOLDER = exports.CONFIG_SRC_FOLDER = exports.CONFIG_DIST_FOLDER = void 0;
const typescript_1 = __importDefault(require("typescript"));
// meaningful defaults
exports.CONFIG_DIST_FOLDER = './dist';
exports.CONFIG_SRC_FOLDER = './src';
exports.CONFIG_PAGES_FOLDER = './src/pages';
exports.CONFIG_PUBLIC_FOLDER = './public';
exports.CONFIG_HOOKS_FOLDER = './src/hooks';
// buildOptions.sitemap
exports.CONFIG_BUILD_SITEMAP = true;
// buildOptions.robotsTxt
exports.CONFIG_BUILD_ROBOTS_TXT = {
    enabled: true,
};
// buildOptions.target
exports.CONFIG_BUILD_TARGET = typescript_1.default.ScriptTarget.ES2018;
// buildOptions.pageUrlFormat
exports.CONFIG_BUILD_PAGE_URL_FORMAT = 'directory';
// devOptions.hostname
exports.CONFIG_DEVOPTIONS_HOSTNAME = 'localhost';
// devOptions.port
exports.CONFIG_DEVOPTIONS_PORT = 3000;
// devOptions.useTls
exports.CONFIG_DEVOPTIONS_USETLS = false;
// devOptions.useOptimizer
exports.CONFIG_DEVOPTIONS_USEOPTIMIZER = true;
