// vanil/dist/src/core/config/transpileOptions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.geTranspileOptions = void 0;
const typescript_1 = __importDefault(require("typescript"));
const defaults_1 = require("./defaults");
/** decides on the default TypeScript compiler options to use */
const geTranspileOptions = (context) => {
    var _a;
    return {
        compilerOptions: {
            module: typescript_1.default.ModuleKind.CommonJS,
            target: ((_a = context.config.buildOptions) === null || _a === void 0 ? void 0 : _a.target) || defaults_1.CONFIG_BUILD_TARGET,
            allowJs: true,
            jsx: typescript_1.default.JsxEmit.React,
            jsxFragmentFactory: 'Vanil.tsx',
            jsxFactory: 'Vanil.tsx',
            skipLibCheck: true,
            checkJs: false,
            skipDefaultLibCheck: true,
        },
    };
};
exports.geTranspileOptions = geTranspileOptions;
